source/CustomSale.cpp

```diff
@@ -0,0 +1,287 @@
+	}
+	if(sellType == SellType::NONE)
+		sellType = SellType::VISIBLE;
+}
+
+
+
+// Can only add between CustomSales of same sellType.
+bool CustomSale::Add(const CustomSale &other)
+{
+	// sellType::NONE means a new CustomSale created with no data and default sellType
+	if(this->sellType == SellType::NONE)
+		this->sellType = other.sellType;
+	else if(other.sellType != this->sellType)
+		return false;
+	for(const auto& it : other.relativePrices)
+	{
+		const auto& item = relativePrices.find(it.first);
+		if(item == relativePrices.cend())
+			relativePrices[it.first] = it.second;
+		else if(item->second < it.second)
+			item->second = it.second;
+	}
+	for(const auto& it : other.relativeOffsets)
+	{
+		const auto& item = relativeOffsets.find(it.first);
+		if(item == relativeOffsets.cend())
+			relativeOffsets[it.first] = it.second;
+		else
+			item->second += it.second;
+	}
+	for(const auto& it : other.relativeOutfitPrices)
+	{
+		const auto& item = relativeOutfitPrices.find(it.first);
+		if(item == relativeOutfitPrices.cend())
+			relativeOutfitPrices[it.first] = it.second;
+		else if(item->second < it.second)
+			item->second = it.second;
+	}
+	for(const auto& it : other.relativeOutfitOffsets)
+	{
+		const auto& item = relativeOutfitOffsets.find(it.first);
+		if(item == relativeOutfitOffsets.cend())
+			relativeOutfitOffsets[it.first] = it.second;
+		else
+			item->second += it.second;
+	}
+	return true;
+}
+
+
+
+double CustomSale::GetRelativeCost(const Outfit *item) const
+{
+	const auto& baseRelative = relativeOutfitPrices.find(item);
+	double baseRelativePrice = (baseRelative != relativeOutfitPrices.cend() ? baseRelative->second : DEFAULT);
+	if(baseRelativePrice == DEFAULT)
+		for(const auto& it : relativePrices)
+			if(it.first->Has(item))
+			{
+				baseRelativePrice = it.second;
+				break;
+			}
+	const auto& baseOffset = relativeOutfitOffsets.find(item);
+	double baseOffsetPrice = (baseOffset != relativeOutfitOffsets.cend() ? baseOffset->second : DEFAULT);
+	for(const auto& it : relativeOffsets)
+		if(it.first->Has(item))
+		{
+			if(baseOffsetPrice == DEFAULT)
+				baseOffsetPrice = 0.;
+			baseOffsetPrice += it.second;
+		}
+	if(baseRelativePrice != DEFAULT)
+		return baseRelativePrice + (baseOffsetPrice != DEFAULT ? baseRelativePrice * baseOffsetPrice : 0.);
+	else if(baseOffsetPrice != DEFAULT)
+		return 1. + baseOffsetPrice;
+	else
+		return -1.;
+}
+
+
+
+CustomSale::SellType CustomSale::GetSellType() const
+{
+	return sellType;
+}
+
+
+
+const std::string &CustomSale::GetShown(CustomSale::SellType sellType)
+{
+	return show.find(sellType)->second;
+}
+
+
+
+const Sale<Outfit> &CustomSale::GetOutfits() const
+{
+	seen.clear();
+	for(auto it : relativeOutfitPrices)
+		seen.insert(it.first);
+	for(auto it : relativeOutfitOffsets)
+		seen.insert(it.first);	
+	for(auto sale : relativePrices)
+		seen.Add(*sale.first);
+	for(auto sale : relativeOffsets)
+		seen.Add(*sale.first);	
+	return seen;
```

Comment:
> If you're going to always clear and return rather than memoizing the value, consider removing the class field and instead returning by value.

The thing is this method is only used once when initiating the MapOutfitterPanel, meaning it would be pointless to memorize it.
So I will indeed just return by value.